llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp

```diff
@@ -4309,7 +4312,18 @@ bool AMDGPUAsmParser::validateVOPLiteral(const MCInst &Inst,
       continue;
 
     if (MO.isImm() && !isInlineConstant(Inst, OpIdx)) {
-      uint32_t Value = static_cast<uint32_t>(MO.getImm());
+      uint64_t Value = static_cast<uint64_t>(MO.getImm());
+      bool IsFP = AMDGPU::isSISrcFPOperand(Desc, OpIdx);
+      bool IsValid32Op = AMDGPU::isValid32BitLiteral(Value, IsFP);
+
+      if (!IsValid32Op && !isInt<32>(Value) && !isUInt<32>(Value)) {
+        Error(getLitLoc(Operands), "invalid operand for instruction");
+        return false;
+      }
+
+      if (IsFP && IsValid32Op)
+        Value = Hi_32(Value);
```

Comment:
I think I have finally understood this code. I would prefer to write it as:
```suggestion
      bool IsFP64 = AMDGPU::isSISrcFPOperand(Desc, OpIdx) && AMDGPU::getOperandSize(Desc.operands()[OpIdx]) == 8;

      if (!AMDGPU::isValid32BitLiteral(Value, IsFP64)) {
        Error(getLitLoc(Operands), "invalid operand for instruction");
        return false;
      }

      if (IsFP64)
        Value = Hi_32(Value);
```
(The call to `getOperandSize` is still pretty ugly. Maybe there is a better way of detecting fp64 operands?)

As a corolloray, maybe change the name of `isValid32BitLiteral`'s second argument from `IsFP` to `IsFP64`? Or maybe even remove that helper function altogether and inline it here? The only other use of it is in an assert where it can just be replaced by `assert(Lo_32(Imm) == 0)`.

Reply:
I need this helper function in at least 3 different patches downstream, one is even approved by you. I am literally stuck without it. I can inline it here but it will be reintroduced anyway.

Yes, I do not like a use of getOperandSize when I can avoid it.

I also do not like changing isValid32BitLiteral's argument to IsFP64 as it brings even more context than it is needed. In fact this is encoding problems, in an ideal world it should not exist at all. I've seen fp32 as a literal to an fp64 instruction and vice versa. This just wraps my mind as I do not understand what does that mean. I mean I really want to keep it simple.

Reply:
> I also do not like changing isValid32BitLiteral's argument to IsFP64 as it brings even more context than it is needed.

The problem with calling it `IsFP` is that if I write this:
```
  bool IsFP = AMDGPU::isSISrcFPOperand(Desc, OpIdx);
  if (!AMDGPU::isValid32BitLiteral(Value, IsFP))
    Error(getLitLoc(Operands), "invalid operand for instruction");
```
then all 16-bit FP operands are rejected. (Maybe all 32-bit ones too, I didn't check.)